python/cuml/metrics/pairwise_kernels.py

```diff
@@ -0,0 +1,307 @@
+#
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import math
+import inspect
+from numba import cuda
+import cupy as cp
+import numpy as np
+import cuml.internals
+from cuml.metrics import pairwise_distances
+
+
+def linear_kernel(X, Y):
+    return cp.dot(X, Y.T)
+
+
+def polynomial_kernel(X, Y, degree=3, gamma=None, coef0=1):
+    if gamma is None:
+        gamma = 1.0 / X.shape[1]
+    K = cp.dot(X, Y.T)
+    K *= gamma
+    K += coef0
+    K **= degree
+    return K
+
+
+def sigmoid_kernel(X, Y, gamma=None, coef0=1):
+    if gamma is None:
+        gamma = 1.0 / X.shape[1]
+
+    K = cp.dot(X, Y.T)
+    K *= gamma
+    K += coef0
+    cp.tanh(K, K)
+    return K
+
+
+def rbf_kernel(X, Y, gamma=None):
+    if gamma is None:
+        gamma = 1.0 / X.shape[1]
+
+    K = cp.asarray(pairwise_distances(X, Y, metric='sqeuclidean'))
+    K *= -gamma
+    cp.exp(K, K)
+    return K
+
+
+def laplacian_kernel(X, Y, gamma=None):
+    if gamma is None:
+        gamma = 1.0 / X.shape[1]
+
+    K = -gamma * cp.asarray(pairwise_distances(X, Y, metric='manhattan'))
+    cp.exp(K, K)
+    return K
+
+
+@cuda.jit(device=True)
+def cosine_similarity_element(x, y):
+    x_norm = 0.0
+    y_norm = 0.0
+    z = 0.0
+    for i in range(len(x)):
+        z += x[i]*y[i]
+        x_norm += x[i] * x[i]
+        y_norm += y[i] * y[i]
+    return z / math.sqrt(x_norm * y_norm)
```

Comment:
I believe this is how the `pairwise_distances` are computing the cosine as well (with exception that it's the ` 2  - [a.dot(b) / (sqrt(x_l2_norm) * sqrt(y_l2_norm)]` (and the sqrt should distributive).

Reply:
The sklearn version here (https://github.com/scikit-learn/scikit-learn/blob/9f85c9d44965b764f40169ef2917e5f7a798684f/sklearn/metrics/pairwise.py#L1265), when ported using cupy and using cumls normalize function, seemed to be numerically unstable to me. I can look more into it if necessary.

Reply:
I'm just wondering if correcting the cosine distance from `cuml.metric.pairwise_distances` back to a similarity might help eliminate the jit overhead from this one as well. If not, we can always look further into it in the future. Thanks for changing the other ones!

Reply:
Ah, I see. I have changed that one to use cosine distance too.